Interrogator.xUnit/NotConcurrentAttribute.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Interrogator.xUnit.Common;
+using Interrogator.xUnit.Utilities;
+
+namespace Interrogator.xUnit
+{
+	[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+	public class NotConcurrentAttribute : DependsOnAttribute
+	{
+		public enum ConcurrencyScope
+		{
+			Class,
+			Namespace,
+			Assembly
+		}
+
+		public string GroupName { get; }
+		public ConcurrencyScope Scope { get; }
+		public string CurrentMethodName { get; }
+
+		public NotConcurrentAttribute(string groupName, [CallerMemberName]string currentMethodName = null) : this(groupName, ConcurrencyScope.Class, currentMethodName) { }
+
+		public NotConcurrentAttribute(string groupName, ConcurrencyScope scope, [CallerMemberName]string currentMethodName = null) : base(null)
+		{
+			GroupName = $"{groupName}_{scope}"; // Adding scope to group name ensures that different scoped attributes don't collide
+			Scope = scope;
+			CurrentMethodName = currentMethodName;
+		}
+
+		internal override Result<Option<MethodInfo>, string> TryGetMethod(Type containingType)
+		{
+			var result = (Type ?? containingType)
+				.TryGetMethod(CurrentMethodName, ParameterTypes)
+				.Select(currentMethod => GetPreviousInChain(currentMethod, containingType, Scope, GroupName));
+			return result;
+		}
+
+		private static Option<MethodInfo> GetPreviousInChain(MethodInfo currentMethod, Type type, ConcurrencyScope scope, string groupName)
+		{
+			var group = GetSameGroupMethods(scope, type, groupName);
+			return GetPreviousMethod(currentMethod, group);
+		}
+
+		private static MethodInfo[] GetSameGroupMethods(ConcurrencyScope scope, Type type, string groupName)
+		{
```

Comment:
This sounds expensive. Won't this happen once for every method marked up with the NotConcurrent attribute. This could mean walking all types in the assembly hundreds of times. Maybe it isn't actually a big deal.

Reply:
Maybe just a static dictionary to accelerate these lookups? In particular, the assembly search could be expensive. Maybe just build the dictionaries in a static constructor?